### PR TITLE
Port SRW App to WCOSS2

### PR DESCRIPTION
**Pull request by chan-hoo**
## DESCRIPTION OF CHANGES: 
- Port UFS SRW App to WCOSS2
- Since HPC-stack is not available on WCOSS2 yet, this is a temporary module file.
- work with the latest hashes of UFS_UTILS and UPP (shown in PR #266)

## TESTS CONDUCTED: 
- Tested hashes:
UFS_UTILS: b6efa86
ufs weather model: 96dffa1
UPP: fbd41a5

- Build steps:
1. git clone -b feature/wcoss2 https://github.com/chan-hoo/ufs-srweather-app
2. cd ufs-srweather-app
3. ./manage_externals/checkout_externals
4. ./devbuild.sh --platform=wcoss2

## ISSUE: 
Fixes issue mentioned in #249

## CONTRIBUTORS: 
@BenjaminBlake-NOAA

### CODE REVIEW

File: Externals.cfg
Comment (JeffBeck-NOAA) [inline Externals.cfg:35]: Have these hashes of UFS_UTILS and UPP been tested with the SRW App on other platforms than WCOSS2?

Comment (chan-hoo) [reply Externals.cfg:35]: @JeffBeck-NOAA, this hash has been tested with PR #266 on other machines. Once this PR is approved, I'll change this hash to the original in this PR. This hash is just for testing.

**Review by RatkoVasic-NOAA — APPROVED** [review]

File: Externals.cfg
Comment (RatkoVasic-NOAA) [reply Externals.cfg:35]: Tested on wcoss2. Build was successful.

Comment (JeffBeck-NOAA) [reply Externals.cfg:35]: What is the status of these hashes on Cheyenne?  @mkavulich, @mark-a-potts, do we need to update hpc-stack or otherwise install newer libraries for this PR and for PR #266?

File: devbuild.sh
Comment (danielabdi-noaa) [inline devbuild.sh]: I think it is better to leave this to the user. This looks like we are going back to auto-detecting machine just like `detect_machine.sh` used to do. If the user is on 'cactus/dogwood' and wants to use wcoss2 modulefiles, he can just pretend to be on wcoss2 and pass `-p=wcoss2` instead.

File: etc/lmod-setup.sh
Comment (danielabdi-noaa) [inline etc/lmod-setup.sh]: Can this go to the modulefiles instead? Unless it is loading Lmod and it is something that can be loaded in modulefiles, I think it should go there.

File: devbuild.sh
Comment (chan-hoo) [reply devbuild.sh]: @danielabdi-noaa, the wcoss system is weird. It switches between `cactus` and `dogwood` on a regular basis. So both `cactus` and `dogwood` are called `wcoss2`. This if-statement prevents a user on wcoss2, either on cactus or on dogwood, from looking into this script or typing --help to figure out what is wrong.

File: etc/lmod-setup.sh
Comment (chan-hoo) [reply etc/lmod-setup.sh]: @danielabdi-noaa, when a user logs in wcoss2, the module `envvar/1.0` is automatically loaded by default. In the wcoss2 user wiki, the following is described in the section of "Lmod and User Environments": `Please note that when doing a module purge you need to reload envvar, module load envvar/1.0, in your environment (interactive or batch) to get the appropriate default modules. NCO has implemented the standard of doing a “module reset”. This implements the module purge and module load of envvar/1.0 in one step.` So I think this command would be better to be placed here. What do you think about this?

File: devbuild.sh
Comment (danielabdi-noaa) [reply devbuild.sh]: @chan-hoo But then you would have to make the same changes in `test/build.sh` as well. I still think it is better to keep it simple and let the user explicitly specify he wants to use wcoss2 modulefiles for the build. 
I was about to suggest that instead of changing `devbuild.sh` and `test/build.sh`, to provide modulefiles for `cactus/dogwood` that are sym-linked to `wcoss2` ones, but there is lots of code specific for wcoss2 `if [ platform = "wcoss" ]` that it probably complicates the code even more ...

Comment (chan-hoo) [reply devbuild.sh]: @danielabdi-noaa, I understand what your concern is. We might have a different viewpoint. I stand on a user's point of view, while you are on a developer's point of view :) :) If you feel uncomfortable about this statement, I'll remove it.

File: etc/lmod-setup.sh
Comment (danielabdi-noaa) [reply etc/lmod-setup.sh]: I think `module load envar/1.0` can go to the top of `build_wcoss2_*`. Have you tested this change and found it to cause a problem? It looks like it is loading default modules which should not be any different from loading custom modulefiles of SRW. It is not changing the module app, as is the case for example on `gaea/odin` going from Cray->Lmod or on macos where you have no Lmod when you start. Also, the way you have it now if the user does a regular `module purge` on the command line, he has to follow it up with `module load envvar/1.0`, which would be avoided if you have the latter in `build_wcoss2`. But for this to be the case, you have to test  loading envvar in `build_wcoss2` works in the first place ...

File: devbuild.sh
Comment (christopherwharrop-noaa) [reply devbuild.sh]: @chan-hoo - I agree with @danielabdi-noaa, the user needs to be responsible for knowing which machine they are on and specify the correct machine.  The use of if-then-else clauses and checking for specific machines is an anti-pattern that leads to overly complicated hard-to-understand and unmaintainable code.  The names of machines should not appear anywhere in the UFS code or scripts, and I will not approve of changes that includes that.  Please remove this check.

File: etc/lmod-setup.sh
Comment (christopherwharrop-noaa) [reply etc/lmod-setup.sh]: Once again, machine specific things need to be isolated to the machine-specific configuration.  I agree with @danielabdi-noaa. This belongs in the wcoss configuration module file.  It doesn't make sense here, and pollutes the code with machine-specific things in places users and developers don't expect.

**Review by christopherwharrop-noaa — CHANGES_REQUESTED** [review]

File: etc/lmod-setup.sh
Comment (chan-hoo) [reply etc/lmod-setup.sh]: @danielabdi-noaa, okay, I'll test it and move it to `build_wcoss2_*` unless it cases any issues.

Comment (chan-hoo) [reply etc/lmod-setup.sh]: @danielabdi-noaa , @christopherwharrop-noaa , what do you think of replacing "module purge and module load envvar" with "module reset"?

File: devbuild.sh
Comment (chan-hoo) [reply devbuild.sh]: @christopherwharrop-noaa, okay, no problem. I'll remove it. My concern is that all users are not as smart as you expect :) :)

File: etc/lmod-setup.sh
Comment (christopherwharrop-noaa) [reply etc/lmod-setup.sh]: If you are putting it in the `build_wcoss*` modulefile, then you can do whatever you want there as that is completely isolated behavior for that particular machine configuration.  My objection is the test for machine name in a non-machine-specific config/module file.

Comment (chan-hoo) [reply etc/lmod-setup.sh]: Do you mean you don't like `module reset`, either? It is an official nco standard on wcoss2.

File: devbuild.sh
Comment (christopherwharrop-noaa) [reply devbuild.sh]: @chan-hoo - Thank you for making that change.  One lesson I've learned the hard way is that making things stupid-proof is a fool's errand.  It's just not possible and causes more problems than it solves.  Knowing what machine you are using is a reasonable expectation.

File: etc/lmod-setup.csh
Comment (christopherwharrop-noaa) [inline etc/lmod-setup.csh]: @chan-hoo - This file is problematic from a design point of view because it is unscalable/unmaintainable/unportable/untestable.  This stuff should be moved to some kind of individual machine-specific configuration file.  But, since you are just following the existing template, I can't argue with this change right now.  But this should be flagged as something that needs rethinking with respect to the overall design.

**Review by christopherwharrop-noaa — APPROVED** [review]

File: etc/lmod-setup.csh
Comment (chan-hoo) [reply etc/lmod-setup.csh]: @christopherwharrop-noaa, thank you for the comment. If you still think that it is better to separate two module commands and move envvar to build_wcoss2_intel, please let me know. I'll talk about this with my EMC colleagues.

File: etc/lmod-setup.sh
Comment (christopherwharrop-noaa) [reply etc/lmod-setup.sh]: There are really two problems.  First, manipulation of modules really belongs in the machine-specific modulefiles.  The other big problem is the `elif [ "$L_MACHINE" = wcoss2 ]`.  That is bad design and should be fixed.  The conditional tests in this file for which machine you are on needs to be removed and moved from the code/scripts to machine-specific configurations.  But, it appears that you are just following the existing template here, and so I can approve this as is for now because fixing this design problem is beyond the scope of your proposed changes.

**Review by dshawul — APPROVED** [review]

**Review by danielabdi-noaa — APPROVED** [review]

File: etc/lmod-setup.sh
Comment (chan-hoo) [reply etc/lmod-setup.sh]: @christopherwharrop-noaa, now I got what your point was. You pointed out that the if-statement itself in the 'lmod-setup.sh' file would cause the issues you mentioned. Is this correct? I thought that you were talking about the command 'module load envvar'. I'll remove 'wcoss2' from the if-statement. Sorry for my late grasp! :) :)

Comment (danielabdi-noaa) [reply etc/lmod-setup.sh]: @chan-hoo I think it is Ok both ways. It is desirable if sourcing `etc/lmod-setup.sh/csh` can be avoided, one less step. On most systems that is actually the case and all you get if you do source it is the `module purge` in the `else:` condition.  In my opinion, those two etc files should be restricted to setting up Lmod, and loading default modules (envvar) can be considered part of setting up Lmod. If it was upto me, I would rather have the user load Lmod manually before building SRW :), but I know your stand on this. Thanks for addressing the requests.

**Comment by natalie-perlin** [comment]
@gspetro-NOAA - is WCOSS2 being supported as Level 1 systems? It seems like WCOSS2 has been removed from the supported platforms, following the discussion on SRW App meeting on May 24, 2022 (but I may be wrong).

**Comment by chan-hoo** [comment]
@natalie-perlin, WCOSS1 (wcoss_dell_p3) will go away in June. So it will be removed from SRW App soon (maybe in July). Instead, WCOSS2 (cactus/dogwood) will be on the list of the tier-1 platforms.

**Comment by mark-a-potts** [comment]
> @natalie-perlin, WCOSS1 (wcoss_dell_p3) will go away in June. So it will be removed from SRW App soon (maybe in July). Instead, WCOSS2 (cactus/dogwood) will be on the list of the tier-1 platforms.

For the release, I think we should leave WCOSS2 off. It is not yet operational and I am not sure we have all the required conda packages installed there, yet. If you like, you can add a "coming soon..." for it, though.

**Comment by chan-hoo** [comment]
@mark-a-potts, yes. I agree. Both WCOSS1 (wcoss_dell_p3) and WCOSS2 will not be available on this release version.

**Review by BenjaminBlake-NOAA — APPROVED** [review]
I was able to build the codes successfully on WCOSS2.

**Pull request merged**
